### PR TITLE
Merge jaspResults into jaspBase

### PR DESCRIPTION
**Pull request by vandenman**
Fixes https://github.com/jasp-stats/INTERNAL-jasp/issues/1888 and supersedes https://github.com/jasp-stats/jasp-desktop/pull/4734

This will obviously need a lot of testing, I'm not 100% sure how to approach that. I also made some minor changes to the .cpp and .h files that used to be in jaspResults and to some R files (mainly cleaning up stuff).

@JorisGoosen any comments?

### CODE REVIEW

File: Engine/CMakeLists.txt [hunk 1]
```diff
@@ -15,8 +15,8 @@ if(LINUX)
 
 endif()
 
-file(GLOB_RECURSE HEADER_FILES "${CMAKE_CURRENT_LIST_DIR}/*.h")
-file(GLOB_RECURSE SOURCE_FILES "${CMAKE_CURRENT_LIST_DIR}/*.cpp")
+file(GLOB HEADER_FILES "${CMAKE_CURRENT_LIST_DIR}/*.h")
```
Comment (JorisGoosen) [inline Engine/CMakeLists.txt:18]: why not recurse?

File: Modules/install-jaspBase.R.in [hunk 2]
```diff
@@ -19,7 +19,7 @@ if (Sys.info()["sysname"] == "Darwin" || Sys.info()["sysname"] == "Windows") {
 computeHash <- function(modulePkg) {
   srcFiles <- c(
     list.files(modulePkg,                    recursive=TRUE, full.names = TRUE, pattern = '(NAMESPACE|DESCRIPTION)$'),
-    # list.files(file.path(modulePkg, 'src'),  recursive=TRUE, full.names = TRUE, pattern = '(\\.(cpp|c|hpp|h)|(Makevars|Makevars\\.win))$'),
+    list.files(file.path(modulePkg, 'src'),  recursive=TRUE, full.names = TRUE, pattern = '(\\.(cpp|c|hpp|h)|(Makevars|Makevars\\.win))$'),
```
Comment (JorisGoosen) [inline Modules/install-jaspBase.R.in]: maybe we should add `Makevars.ucrt` as well? 
https://developer.r-project.org/Blog/public/2021/03/12/windows/utf-8-toolchain-and-cran-package-checks/

> R now understands .ucrt suffix of Makevars and Makefile files.

File: R-Interface/CMakeLists.txt [hunk 3]
```diff
@@ -28,10 +28,8 @@
 
 list(APPEND CMAKE_MESSAGE_CONTEXT R-Interface)
 
-file(GLOB SOURCE_FILES "${CMAKE_CURRENT_LIST_DIR}/jaspResults/src/*.cpp")
 list(APPEND SOURCE_FILES "${CMAKE_CURRENT_LIST_DIR}/jasprcpp.cpp")
 
-file(GLOB HEADER_FILES "${CMAKE_CURRENT_LIST_DIR}/jaspResults/src/*.h")
```
Comment (JorisGoosen) [inline R-Interface/CMakeLists.txt:44]: Aah I guess it was not recurse here already, fine then.

File: R-Interface/CMakeLists.txt [hunk 4]
```diff
@@ -28,10 +28,8 @@
 
 list(APPEND CMAKE_MESSAGE_CONTEXT R-Interface)
 
-file(GLOB SOURCE_FILES "${CMAKE_CURRENT_LIST_DIR}/jaspResults/src/*.cpp")
 list(APPEND SOURCE_FILES "${CMAKE_CURRENT_LIST_DIR}/jasprcpp.cpp")
```
Comment (JorisGoosen) [inline R-Interface/CMakeLists.txt:41]: But the APPEND here should be GLOB just like at line 33

File: Engine/CMakeLists.txt (hunk 1 above)
Comment (JorisGoosen) [reply Engine/CMakeLists.txt:18]: Well for Engine it doesnt really matter too much except that if you *do* add it it might add the sources/headers from `jaspResults` to qtcreator (which is nice).
Although they shouldnt be added to `add_executable`.

Instead maybe you could try adding:
```
file(GLOB JASPRESULTS_HEADER_FILES "${CMAKE_CURRENT_LIST_DIR}/jaspBase/src/*.h")
file(GLOB JASPRESULTS_SOURCE_FILES "${CMAKE_CURRENT_LIST_DIR}/jaspBase/src/*.cpp")
```


We could also think about moving the headers to `/inst/include/` (which would necessitate changes to the above) but that would make it possible (in the distant future I guess) for others to extend "jaspResults" and at least to interface on it?
On the other hand this might not be the right time for that.

File: R-Interface/jasprcpp.cpp [hunk 5]
```diff
@@ -157,38 +162,21 @@ void STDCALL jaspRCPP_init(const char* buildYear, const char* version, RBridgeCa
...
-	jaspRCPP_parseEvalQNT("source(file='../Modules/writeImage.R')");
-	jaspRCPP_parseEvalQNT("source(file='../Modules/zzzWrappers.R')");
-	jaspRCPP_parseEvalQNT("source(file='../Modules/workarounds.R')");
-#endif
+//	jaspRCPP_logString("Initializing jaspBase.\n");
+//	jaspRCPP_parseEvalQNT("library(jaspBase)");
 
-	jaspRCPP_logString("initEnvironment().\n");
-	jaspRCPP_parseEvalQNT("initEnvironment()");
+//	if we have a separate engine for each module then we should move these kind of hacks to the .onAttach() of each module (instead of loading BayesFactor when Descriptives is requested).
```
Comment (JorisGoosen) [inline R-Interface/jasprcpp.cpp:185]: for `development` that is the case but I see no reason why this wouldnt work with `stable` either, `jaspBase` would only be loaded once right?

And if you forcibly `library(jaspBase)` here somewhere (for now anyway) it would load the one from the bundled main R library, no?

File: R-Interface/jasprcpp.cpp [hunk 6]
```diff
@@ -70,6 +69,12 @@ static std::string				_R_HOME = "";
 
 bool shouldCrashSoon = false; //Simply here to allow a developer to force a crash
 
+std::string jaspNativeToUtf8(const Rcpp::String & in)
```
Comment (JorisGoosen) [inline R-Interface/jasprcpp.cpp:72]: Why is this here? Doesnt this work differently from the ones in `jaspObject`?
It doesn't matter too much mind you as we can throw it all away soon I hope (after the R 4.2.1 PR is merged) but I just don't understand why you didnt copy both and entirely:
```
std::string jaspNativeToUtf8(const Rcpp::RObject &in)
{
	if(in.isNULL())
		return "";
	
	return jaspNativeToUtf8(Rcpp::as<Rcpp::String>(in));
}


std::string jaspNativeToUtf8(const Rcpp::String & in)
{
#ifdef _WIN32
	#ifdef JASP_R_INTERFACE_LIBRARY
		return jaspRCPP_nativeToUtf8(in);
	#else
		return in; //If running in R then it's the users problem really.
	#endif	
#else
	return in;
#endif
}
```
Although I guess the object one might not be necessary in here

File: R-Interface/jasprcpp.cpp (hunk 5 above)
Comment (vandenman) [reply R-Interface/jasprcpp.cpp:185]: well, sure but if you look at https://github.com/jasp-stats/jaspBase/blob/5d79025af56f6c26788302af0d38cb2b780e001f/R/common.R#L157-L175 then it means that for every module we're going to do `try(loadNameSpace("BayesFactor"))`. Most modules don't depend on BayesFactor, so this just takes up unnecessary time. For the modules where this matters we can move this to `.onLoad`. 

The other two things should just move to `runJaspResults` as far as I'm concerned (`registerFonts()` is already called there anyway).

> And if you forcibly library(jaspBase) here somewhere (for now anyway) it would load the one from the bundled main R library, no?

It would load the first one it finds in the libPaths. I think the main R library is the last one in the libPaths, so it's probably not loading that one.

File: R-Interface/jasprcpp.cpp (hunk 6 above)
Comment (vandenman) [reply R-Interface/jasprcpp.cpp:72]: I remember not caring about windows when I tried to make this work the previous time and this time I didn't bother to change it :) I can remove all calls to this function and the definition itself if you think that's better?

File: R-Interface/CMakeLists.txt (hunk 4 above)
Comment (vandenman) [reply R-Interface/CMakeLists.txt:41]: isn't it `list(APPEND ...)` and `FILE(GLOB ...)`? Cmake throws an error if I do `list(GLOB ... )`

File: Engine/CMakeLists.txt (hunk 1 above)
Comment (vandenman) [reply Engine/CMakeLists.txt:18]: > why not recurse?

because it got added to `add_executable` and then compilation failed. Moving things to `/inst/include/` is a good idea, but we can do that at a later stage.

> for Engine it doesnt really matter too much except that if you do add it it might add the sources/headers from jaspResults to qtcreator (which is nice).

yeah that would be nice. This now works with 

```cmake
file(GLOB JASPRESULTS_HEADER_FILES "${CMAKE_CURRENT_LIST_DIR}/jaspBase/src/*.h")
file(GLOB JASPRESULTS_SOURCE_FILES "${CMAKE_CURRENT_LIST_DIR}/jaspBase/src/*.cpp")

# this target is not compiled and only exists so that the sources and headers of jaspResults show up in the project and can be opened with Ctrl+k
add_custom_target(jaspResults ALL SOURCES ${JASPRESULTS_SOURCE_FILES} ${JASPRESULTS_HEADER_FILES})
```
image

File: Modules/install-jaspBase.R.in (hunk 2 above)
Comment (vandenman) [reply Modules/install-jaspBase.R.in]: Good idea!

File: R-Interface/jasprcpp.cpp (hunk 5 above)
Comment (JorisGoosen) [reply R-Interface/jasprcpp.cpp:185]: >It would load the first one it finds in the libPaths. I think the main R library is the last one in the libPaths, so it's probably not loading that one.

But this code is run on engine initialization so it doesnt yet get the libpaths for a module. And thus loads the one from mainlibrary.

The rest of your comment sounds fine

File: Engine/CMakeLists.txt (hunk 1 above)
Comment (JorisGoosen) [reply Engine/CMakeLists.txt:18]: aah nice yeah that looks helpful

File: R-Interface/CMakeLists.txt (hunk 4 above)
Comment (JorisGoosen) [reply R-Interface/CMakeLists.txt:41]: I hadn't noticed that it was also a different function, but I interpreted "APPEND" as needing a list to exist already.
However if it works it works

File: R-Interface/jasprcpp.cpp (hunk 6 above)
Comment (vandenman) [reply R-Interface/jasprcpp.cpp:72]: I implemented your suggestion but skipped the one for `RObject` since that's not used here.

Comment (JorisGoosen) [reply R-Interface/jasprcpp.cpp:72]: I thought you were removing them? 
This is also fine btw just curious

Comment (vandenman) [reply R-Interface/jasprcpp.cpp:72]: I'm not sure about windows. If people ever want to build Jasp again with R <4.2.0 then they'd need this right? Also, it seemed like the easiest way to minimize the changes.

**Review by JorisGoosen — APPROVED** [review]

File: R-Interface/jasprcpp.cpp (hunk 6 above)
Comment (JorisGoosen) [reply R-Interface/jasprcpp.cpp:72]: well, yeah but they could also just get an older version of JASP then... I dont really see why that would be useful (or useful enough to keep the old code in just for that).

However, I still need to doublecheck the windows encoding stuff anyway so Ill look at this then too.
